File: src/LooperAnalyzer/Analysis/SymbolUtils.cs
using Microsoft.CodeAnalysis;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace LooperAnalyzer.Analysis
{
    internal static class SymbolUtils
    {
        private static HashSet<IMethodSymbol> _whitelist;
        private static ITypeSymbol _genericIEnumerableType;

        public static void InitializeFromCompilation(Microsoft.CodeAnalysis.Compilation compilation)
        {
            if (_whitelist != null && _genericIEnumerableType != null) return;

            var whitelistNames = new HashSet<string>
            {
                nameof(Enumerable.First),
                nameof(Enumerable.FirstOrDefault),
                nameof(Enumerable.Single),
                nameof(Enumerable.SingleOrDefault),
                nameof(Enumerable.Any),
                nameof(Enumerable.Average),
                nameof(Enumerable.Count),
                nameof(Enumerable.ElementAt),
                nameof(Enumerable.ElementAtOrDefault),
                nameof(Enumerable.Last),
                nameof(Enumerable.LastOrDefault),
                nameof(Enumerable.Max),
                nameof(Enumerable.Min),
                nameof(Enumerable.Single),
                nameof(Enumerable.SingleOrDefault),
                nameof(Enumerable.Sum),
                nameof(Enumerable.ToArray),
                nameof(Enumerable.ToList),
            };

            var filtered = compilation.GetTypeByMetadataName("System.Linq.Enumerable")
                .GetMembers()
                .Where(s => whitelistNames.Contains(s.Name))
                .OfType<IMethodSymbol>();

            _whitelist = new HashSet<IMethodSymbol>(filtered);
            
            _genericIEnumerableType = compilation.GetTypeByMetadataName("System.Collections.Generic.IEnumerable`1");
        }

        public static bool IsOptimizableConsumerMethod(this IMethodSymbol method) => 
            _whitelist.Contains(method.IsExtensionMethod ? method.ReducedFrom : method);

        public static bool IsArrayType(this ITypeSymbol type) => type is IArrayTypeSymbol;

        public static bool IsIEnumerableType(this ITypeSymbol type) =>
            type is INamedTypeSymbol
            && (type.OriginalDefinition == _genericIEnumerableType 
                || type.OriginalDefinition.AllInterfaces.Contains(_genericIEnumerableType));

        public static bool IsOptimizableSourceType(this ITypeSymbol type) => type.IsArrayType() || type.IsIEnumerableType();


    }
}

File: src/Looper.Core/Core/SymbolUtils.fs
module Looper.Core.SymbolUtils

open Microsoft.CodeAnalysis
open System.Collections.Generic

let mutable private allLinqMethods = Unchecked.defaultof<_>
let mutable private genericIEnumerableType = Unchecked.defaultof<_>

let initializeFromCompilation (compilation : Compilation) = 
    // TODO : check if System.Linq is referenced in compilation
    if isNull allLinqMethods || isNull genericIEnumerableType then 
        let methods = 
            compilation.GetTypeByMetadataName("System.Linq.Enumerable").GetMembers() 
            |> Seq.where (fun m -> m :? IMethodSymbol)
        allLinqMethods <- new HashSet<ISymbol>(methods)
        genericIEnumerableType <- compilation.GetTypeByMetadataName("System.Collections.Generic.IEnumerable`1")

let (|LinqMethod|Other|) (symbol : ISymbol) = 
    if isNull allLinqMethods then invalidOp "Not initialized from a compilation"
    match symbol with
    | :? IMethodSymbol as s when s.IsExtensionMethod && allLinqMethods.Contains s.ReducedFrom 
                                 || allLinqMethods.Contains s -> LinqMethod s
    | _ -> Other

let (|ArrayType|_|) (symbol : ISymbol) = 
    match symbol with
    | :? IArrayTypeSymbol as s -> Some s
    | _ -> None

let (|IEnumerableType|_|) (symbol : ISymbol) = 
    match symbol with
    | :? INamedTypeSymbol as s when s.OriginalDefinition = genericIEnumerableType 
                                    || s.OriginalDefinition.AllInterfaces.Contains(genericIEnumerableType) -> Some s
    | _ -> None